imap_processing/swapi/l2/swapi_l2.py

```diff
@@ -12,7 +15,93 @@
 TIME_PER_BIN = 0.167  # seconds
 
 
-def swapi_l2(l1_dataset: xr.Dataset, data_version: str) -> xr.Dataset:
+def solve_full_sweep_energy(
+    esa_lvl5_data: np.ndarray, esa_table_df: pd.DataFrame, lut_notes_df: pd.DataFrame
+) -> npt.NDArray:
+    """
+    Calculate the energy of each full sweep data.
+
+    Parameters
+    ----------
+    esa_lvl5_data : numpy.ndarray
+        The L1 data input.
+    esa_table_df : pandas.DataFrame
+        The ESA unit conversion table that contains first 63 energies.
+    lut_notes_df : pandas.DataFrame
+        The LUT notes table that contains the last 9 fine energies.
+
+    Returns
+    -------
+    energy : numpy.ndarray
+        The energy of each full sweep data.
+    """
+    # Read 0 - 62 energy steps' fixed energy value
+    fixed_energy_values = esa_table_df["Energy"].values[:63]
+
+    # Find last 9 fine energy values of all sweeps data
+    # -------------------------------------------------
+    # First, verify that one value in the LUT-notes table's 'ESA DAC (Hex)' column
+    # exactly matches a value in the esa_lvl5_data.
+    has_exact_match = np.isin(esa_lvl5_data, lut_notes_df["ESA DAC (Hex)"].values)
+    if not np.all(has_exact_match):
+        msg = (
+            "These ESA_LVL5 values not found in lut-notes table: "
+            f"{esa_lvl5_data[np.where(~has_exact_match)[0]]} "
+        )
+        raise ValueError(msg)
+
+    # Find index of 71th energy step for all sweeps data in lut-notes table.
+    # Tried using np.where(np.isin(...)) or df.index[np.isin(...)] to find the index
+    # of each value in esa_lvl5_data within the LUT table. However, these methods
+    # return only the unique matching indices — not one index per input value.
+    # For example, given the input:
+    #   ['12F1', '12F1', '12F1', '12F1']
+    # np.where(np.isin(...)) would return:
+    #   [336]
+    # because it finds that '12F1' exists in the LUT and only returns its position once.
+    # What we actually need is:
+    #   [336, 336, 336, 336]
+    # — one index for *each* occurrence in the input, preserving its shape and order.
+    # Therefore, instead of relying on np.isin or similar, we explicitly use
+    # np.where in a loop to find the index of each value in esa_lvl5_data individually,
+    # ensuring the output array has the same shape as the input.
+
+    last_energy_step_indices = np.array(
+        [
+            np.where(lut_notes_df["ESA DAC (Hex)"].values == val)[0][0]
+            for val in esa_lvl5_data
+        ]
+    )
+    # Use back tracking steps to find all 9 fine energy value indices
+    # Eg. [0, -4, -8, ..., -28, -32]
+    steps = np.arange(0, -4 * 9, -4)
+
+    # Find indices of last 9 fine energy values of all sweeps data
+    fine_energy_indices = last_energy_step_indices[:, None] + steps
+
+    # Per SWAPI instruction, floor every indices that result in negative
```

Comment:
Add a NOTE clarifier here to draw attention to it in the editors if its something to call out.

When you say "floor every indices", during back tracking to zero what does that mean? Are the last nine steps the "back tracking to zero" portion of the sweep? Flooring something generally means just nudge it towards zero not necessarily set it to zero, so I'm a bit unclear with the instructions here.

```suggestion
    # NOTE: Per SWAPI instruction, floor every index that result in negative
```